### DISCUSSION TITLE
Use environment variables

### DISCUSSION THREAD

**Discussion by josemiguelo** [stevearc/overseer.nvim#173] [Q&A] [answered]
Hi!
First up, amazing job on this plugin!

I've just started to play with templates and my use case is this: I use several java versions on my machine, and the `maven` tool to manage my projects. On some projects, I append the java version before executing `maven` like so:
```sh
JAVA_HOME=/home/user/.asdf/installs/java/adoptopenjdk-8.0.362+9 mvn ...
``` 
This lets me use the java version my project needs on every specific case. Now, I'm creating a template to automate this  use case (and then chain it with some other tasks) so my component looks like this 
```lua
return {
 ...
  builder = function(params)
    return {
      cmd = {"mvn" },
      args = { "--version" },
    }
  end,
  env = {
    JAVA_HOME = "/home/user/.asdf/installs/java/adoptopenjdk-8.0.362+9",
  },
  condition = { dir = utils.path },
}
```
but this doesn't set the JAVA_HOME the way I want. (I can see in the output I'm using a different jdk version). I tried several things like
```
return {
  cmd = { "JAVA_HOME=/home/user/.asdf/installs/java/adoptopenjdk-8.0.362+9", "mvn" },
  args = { "--version" },
}
```
or
```
return {
  cmd = { "JAVA_HOME=/home/user/.asdf/installs/java/adoptopenjdk-8.0.362+9 mvn" },
  args = { "--version" },
}
```

but the task just hangs and nothing happens.

Maybe this is not the correct way to pass those strings to the `cmd` field. It should receive the actual executable name and I'm missing things. 

BTW if I don't use the env variable, the task executes successfully.

Can someone kindly shed some light on this issue? :)

**Comment by stevearc** [comment, marked as answer]
If you check the docs for [template definitions](https://github.com/stevearc/overseer.nvim/blob/514a5e1af18b490721836fa19b62ca60761e5b59/doc/guides.md#template-definition), the `env` should be at the same level as `cmd` and `args`. For example:
```lua
return {
 ...
  builder = function(params)
    return {
      cmd = {"mvn" },
      args = { "--version" },
      env = {
        JAVA_HOME = "/home/user/.asdf/installs/java/adoptopenjdk-8.0.362+9",
      },
    }
  end,
  condition = { dir = utils.path },
}
```

**Reply by tony-sol** [reply]
Will this
```lua
JAVA_HOME = os.getenv("JAVA_HOME")
```
work?

**Reply by stevearc** [reply]
It should, but I believe that your current env should be passed through by default so that should be unnecessary.

**Comment by josemiguelo** [comment]
You're totally right. I'm sorry for my misunderstanding. But it seems that even when using the `env` struct correctly, the variable is not set correctly; in fact, maven yells at me saying the there's no more JAVA_HOME set. 

I think it's best to not define the JAVA_HOME variable this way, but to use an external tool that sets the the appropriate JAVA_HOME (because of the way I intended to use the env var: `JAVA_HOME=... mvn --version`)

I use `asdf` as a version manager for java, and for every project I define its jdk version.If you use the cwd correctly, it will pick up the jdk correctly as well.

```lua
return {
      cwd = "some path to your project where the JAVA_HOME is already defined',
      cmd = { "mvn" },
      args = { "--version" },
    }
```

If you **really** need to define env variables **before** the actual binary, just create a wrapper
```bash
#!/bin/bash

cd /your/project/path
JAVA_HOME=/home/user/.asdf/installs/java/adoptopenjdk-8.0.362+9 mvn clean install 
```
